examples/common/text_data.py

```diff
@@ -108,15 +121,52 @@ def _tokenize(self, text_sample):
             max_length = None
         else:
             raise ValueError(f"Got unknown group_method='{self.group_method}'.")
-        return self.tokenizer(text_sample['text'],
-                              truncation=truncation,
-                              padding=padding,
-                              max_length=max_length)
+
+        # The first time we tokenize we need to figure out the special token situation,
+        # as pre-concatenated text will already have special tokens
+        if self.add_special_tokens is None:
+            tokens = self.tokenizer(text_sample['text'],
+                                    truncation=truncation,
+                                    padding=padding,
+                                    max_length=max_length,
+                                    add_special_tokens=True)
+            if self.group_method == 'concat':
+                # True is the default value so this will not change any behavior
+                self.add_special_tokens = True
+            else:
+                # if the 2nd token is BOS or 2nd-to-last-is EOS, the special tokens
```

Comment:
why 2nd and 2nd to last instead of first and last? im probably just forgetting something here...

Reply:
Because if the tokenizer adds a BOS token AND we have used that BOS token as our `sep_text` then we will see BOS in position 1, similar for EOS in position -2

Reply:
I had not considered the case where sep_text is not EOS or BOS. Do you think it's okay to assume it is one of those?

Reply:
OK, switched sep_text to bos_text and eos_text